# My attempt at creating a "only time-based" dataset for predicitons.
Perhaps we can mix and match some of the features from this to the other.

In [1]:
#Imports
import pandas as pd
from collections import Counter
import numpy as np
import xgboost as xgb
import datetime
#Must run pip install holidays
import holidays

#Initial Data
receivals_raw_df = pd.read_csv("./data/kernel/receivals.csv")
purchase_orders_raw_df = pd.read_csv("./data/kernel/purchase_orders.csv")
materials_raw_df = pd.read_csv("./data/extended/materials.csv")
transportation_raw_df = pd.read_csv("./data/extended/transportation.csv")

prediction_map_raw_df = pd.read_csv("./data/prediction_mapping.csv")

# Part 1: Preparing Data.
We will try different structures for our dataframe. 

### Option 1
In line with our predictions being "per-day" basis, we will create a dataframe where the net-weight per rm_id per day is aggregated, and we will also include entries for every day (even if there are 0 receivals that day). This data structure might help our models since it is much more similar to the output we want to predict, However, it is hard to incorporate info such as purchase order info.

### Option 2
Dataframe on a "per-receival" basis. This is less in-line with our output, but let's us keep more information. We can also use our models to make predictions of receivals and later convert to a per-day basis as a workaround.

## 1.1 Data Cleaning.
Some data-cleanup is required before going with either option.

In [44]:
valid_ids = prediction_map_raw_df["rm_id"].unique()
print("Unique IDs: ", len(valid_ids))

receivals_df = receivals_raw_df[receivals_raw_df["rm_id"].isin(valid_ids)]

receivals_df = receivals_df.rename(columns={"date_arrival" : "date"})

receivals_df.date = pd.to_datetime(receivals_df.date, errors="coerce", utc=True)
receivals_df.date = receivals_df.date.dt.tz_localize(None)
receivals_df.date = receivals_df.date.dt.normalize()
#Remove duplicates.
receivals_df = receivals_df.drop_duplicates()

#Remove entries where net weight is undefined.
receivals_df = receivals_df.dropna(subset=["net_weight"])

#Merge with purchase orders
receivals_df = receivals_df.merge(purchase_orders_raw_df, on=['purchase_order_id', 'purchase_order_item_no'], how="left", )
#print(receivals_df.head())
#Correct for different weight units
receivals_df.loc[receivals_df['unit'] == 'PUND', 'net_weight'] *= 0.453592
receivals_df.loc[receivals_df['unit'].isna(), 'unit'] = 'kg'
receivals_df.drop(columns="unit")

#We only include recent years to save computation time
receivals_df = receivals_df.merge(transportation_raw_df, on=['purchase_order_id', 'purchase_order_item_no', 'receival_item_no', 'batch_id'], how="left")
#receivals_df.to_csv("data_prep/receivalsxpurchase")

valid_ids = receivals_df.loc[receivals_df["date"] >= pd.Timestamp(day=1, month=1, year=2020)]["rm_id"].unique()



Unique IDs:  203


MemoryError: Unable to allocate 121. MiB for an array with shape (8, 1981170) and data type float64

### 1.2 Data-preparation: Option 1- Per-day, Per-receival
We can make further preparations to our dataset if we only care about each "day". We drop values that does not translate well to days.

consider!
- instead of dropping, some of these can be translated in some way: purchase orders can for example get "purchase_order_counts" or similar

In [ ]:
#Show what columns are in our dataframe
#simple_cleaned_df.info()

#We will group multiple orders together as we will only include receivals per day. That means we can drop or change some rows.
Option1_cleaning = simple_cleaned_df.drop(columns=(
    ["product_id_x", #Messy when grouping receivals per day. Should ideally include?
    #"purchase_order_id", #Messy when grouping
    #"purchase_order_item_no", #Messy when grouping
    #"receival_item_no", #Could include if summarizing makes sense
    "batch_id", #Very many missing, Could include if summarizing makes sense
    "receival_status", #Instructed that all are valid
    "supplier_id", #Messy when grouping, could count
    "product_id_y", #Should just match first product_id
    "created_date_time", #Unlikely to be relevant, messy when grouping
    "modified_date_time", #Unlikely to be relevant, messy when grouping
    "unit_id", #All should be KG now
    "unit", #All should be KG now
    "status_id", #unclear what this does
    "status", #should be handled in prior code
    "product_version", #Messy when grouping
    #"date_arrival", #Redundant with days_as_datetime
    ]))



## Part 2: Feature Engineering- Preparing new data_frame for per_day, per_id.
We make a new dataframe where for we include each day (roghly 20 years) and each id (203) for a total of 203*~7000 rows. Many of these will be zeros, as there are not deliveries every day, but the "format" will match our desired output.

### 2.1 Aggregates:
With the suggested format, we group all receivals per day.

In [ ]:
years = list(range(2019, 2026 + 1))
all_holidays = holidays.Norway(years = years)
all_holidays = list(all_holidays.keys())
all_holidays = pd.to_datetime(all_holidays)
all_holidays = pd.DataFrame({"holidays" : all_holidays})

full_range = pd.date_range(start=pd.Timestamp(day=1, month=1, year=2020), end=pd.Timestamp(day=31, month=5, year=2025))

df_for_all_ids = pd.DataFrame({"datetime" : full_range})
df_for_all_ids["datetime"] = pd.to_datetime(df_for_all_ids["datetime"]).dt.tz_localize(None)
df_for_all_ids["datetime"] = df_for_all_ids["datetime"].dt.normalize()
df_for_all_ids["date"] = df_for_all_ids["datetime"].dt.date
#We should get what day of the week it is, since this is likely a good predictor
df_for_all_ids["day_of_week"] = df_for_all_ids["datetime"].dt.dayofweek
#Weekend is perhaps even stronger predictor: So we create a seperate column for this
#df_for_all_ids["is_weekend"]  = df_for_all_ids['day_of_week'].isin([5, 6])
#df_for_all_ids["is_weekend"] = df_for_all_ids["is_weekend"].astype(int)
df_for_all_ids["month"] = df_for_all_ids["datetime"].dt.month

df_for_all_ids["is_holiday"] = df_for_all_ids["datetime"].isin(all_holidays["holidays"])
df_for_all_ids["is_holiday"] = df_for_all_ids["is_holiday"].astype(int)
print(df_for_all_ids.tail())

df = []
for id in valid_ids:
    id_df = df_for_all_ids
    id_df.rm_id = id

    df.append(id_df)


    

       datetime        date  day_of_week  month  is_holiday
1973 2025-05-27  2025-05-27            1      5           0
1974 2025-05-28  2025-05-28            2      5           0
1975 2025-05-29  2025-05-29            3      5           1
1976 2025-05-30  2025-05-30            4      5           0
1977 2025-05-31  2025-05-31            5      5           0


In [ ]:
processed_groups = []

for (id, date), group in df.groupby(['rm_id', 'date']):
    daily_df = pd.Series({"daily_weight" : group.net_weight.sum()})
    daily_df.id = id
    daily_df.date = date
    daily_df.num_orders = group.purchase_order_id.count()
    daily_df.avg_orders = group.purchase_order_id.mean()
    daily_df.max_orders = group.purchase_order_id.max()
    daily_df.std_orders = group.purchase_order_id.std()
    processed_groups.append(daily_df)

df = pd.concat(processed_groups)

print(df.head())

daily_weight     1152.0
daily_weight     1152.0
daily_weight     2000.0
daily_weight    10000.0
daily_weight    10000.0
dtype: float64


In [ ]:
processed_groups = []

for id, group in df.groupby(['rm_id', 'date']):
    


daily_df = df.groupby(['rm_id', 'date']).agg(
    #The total net_weight for that material on that day
    daily_weight=('net_weight', 'sum'),
    #Number of orders
    num_orders=('purchase_order_id', 'count'),
    average_orders=('net_weight', 'mean'),
    max_orders=('net_weight', 'max'),
    std_orders=('net_weight', 'std'),
    #Product information
    num_product_ids = ('supplier_id', 'unique.count')
)
#daily_aggregates = daily_aggregates.fillna({'daily_weight': 0, 'num_orders': 0, 'avg_order_size': 0, 'max_order_size': 0, 'std_order_size': 0}, inplace=True)
daily_df["average_orders"] = daily_df["average_orders"].fillna(0)
daily_df["max_orders"] = daily_df["max_orders"].fillna(0)
daily_df["std_orders"] = daily_df["std_orders"].fillna(0)

recent_years = pd.Timestamp(day=1, month=1, year=2020)

daily_df = daily_df[daily_df.date > recent_years]


## 2.1 "Common" Features
Some features are decided by dates, and not dependent on rm_id. We create these first to avoid doing this in a loop.

In [5]:

#We get all holidays for a relevant year-range:
years = list(range(2004, 2026 + 1))
all_holidays = holidays.Norway(years = years)
all_holidays = list(all_holidays.keys())
all_holidays = pd.to_datetime(all_holidays)
all_holidays = pd.DataFrame({"holidays" : all_holidays})



print("The first day is: " , first_day)
#We start with finding the range of days
full_range = pd.date_range(start=first_day, end=last_day)
full_range = pd.to_datetime(full_range)

#Remember when we make predictions we should present each day in terms of days since first_day

#Some features are the same for all ID's we create them once to not have to do it many times.
#We should have days normalized, and date
df_for_all_ids = pd.DataFrame({"datetime" : full_range})
df_for_all_ids["datetime"] = pd.to_datetime(df_for_all_ids["datetime"]).dt.tz_localize(None)
df_for_all_ids["datetime"] = df_for_all_ids["datetime"].dt.normalize()
df_for_all_ids["date"] = df_for_all_ids["datetime"].dt.date
#We should get what day of the week it is, since this is likely a good predictor
df_for_all_ids["day_of_week"] = df_for_all_ids["datetime"].dt.dayofweek
#Weekend is perhaps even stronger predictor: So we create a seperate column for this
#df_for_all_ids["is_weekend"]  = df_for_all_ids['day_of_week'].isin([5, 6])
#df_for_all_ids["is_weekend"] = df_for_all_ids["is_weekend"].astype(int)
df_for_all_ids["month"] = df_for_all_ids["datetime"].dt.month

df_for_all_ids["is_holiday"] = df_for_all_ids["datetime"].isin(all_holidays["holidays"])
df_for_all_ids["is_holiday"] = df_for_all_ids["is_holiday"].astype(int)
#Verify py printing how many were marked as holidays
print(df_for_all_ids["is_holiday"].sum(), "days were marked as holidays")



The first day is:  2004-06-15 00:00:00
237 days were marked as holidays


## 2.2 Add features per id in a loop
Here we have a loop for our IDs. We add all the "common" features for each ID as well as some ID specific features

In [ ]:
#To make our dataframe, we make an empty list we can append to.
df_full = []


#For each ID we create fields for all days.
#for i in range(0, len(valid_ids[0:5:1])):
for i in valid_ids:
    #We create sub-group dataframes that we will append to the list.
    #We copy data from the "felles" dataframes
    df_i = df_for_all_ids

    #We include the rm_id
    df_i["rm_id"] = i

    #df_i["rm_id"] = valid_ids[i]

    #For days where we actually have receivals, we get those receivals from the daily_aggregates dataframe
    df_i = df_i.merge(daily_aggregates[daily_aggregates['rm_id'] == i], on=['rm_id', "datetime"], how='left')

    #If there are no receivals in a day, we add a daily weight of 0. as well as zero for the other aggregate metrics.
    df_i["daily_weight"] = df_i["daily_weight"].fillna(0)
    df_i["average_orders"] = df_i["average_orders"].fillna(0)
    df_i["max_orders"] = df_i["max_orders"].fillna(0)
    df_i["std_orders"] = df_i["std_orders"].fillna(0)
    df_i["num_orders"] = df_i["num_orders"].fillna(0)

    #We create a column for cumulative weight, although I'm not sure it should be included
    df_i['cumulative_weight'] = df_i['daily_weight'].cumsum()

    #df_i["date_shifted_200"] = df_i["date"].shift(200)
    #We append to the list and proceed to the next id in the loop
    df_full.append(df_i)

#When all entries (per_id_per_loop) is added, we create a dataframe of the list.
df_final = pd.concat(df_full, ignore_index=True)



In [7]:
#Explore data
df_final.info()
df_final.tail()




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1521079 entries, 0 to 1521078
Data columns (total 12 columns):
 #   Column             Non-Null Count    Dtype         
---  ------             --------------    -----         
 0   datetime           1521079 non-null  datetime64[ns]
 1   date               1521079 non-null  object        
 2   day_of_week        1521079 non-null  int32         
 3   month              1521079 non-null  int32         
 4   is_holiday         1521079 non-null  int64         
 5   rm_id              1521079 non-null  int64         
 6   daily_weight       1521079 non-null  float64       
 7   num_orders         1521079 non-null  float64       
 8   average_orders     1521079 non-null  float64       
 9   max_orders         1521079 non-null  float64       
 10  std_orders         1521079 non-null  float64       
 11  cumulative_weight  1521079 non-null  float64       
dtypes: datetime64[ns](1), float64(6), int32(2), int64(2), object(1)
memory usage: 127.7+

,datetime,date,day_of_week,month,is_holiday,rm_id,daily_weight,num_orders,average_orders,max_orders,std_orders,cumulative_weight
1521074,2024-12-15,2024-12-15,6,12,0,4501,0.0,0.0,0.0,0.0,0.0,48180.0
1521075,2024-12-16,2024-12-16,0,12,0,4501,0.0,0.0,0.0,0.0,0.0,48180.0
1521076,2024-12-17,2024-12-17,1,12,0,4501,0.0,0.0,0.0,0.0,0.0,48180.0
1521077,2024-12-18,2024-12-18,2,12,0,4501,0.0,0.0,0.0,0.0,0.0,48180.0
1521078,2024-12-19,2024-12-19,3,12,0,4501,0.0,0.0,0.0,0.0,0.0,48180.0


In [8]:
#Print to file
df_final.to_csv("cleaned_data/cleaned_df_final_test.csv", index=False)